### PR TITLE
Add local versions for global thresholds

### PR DESCRIPTION
**Pull request by stelfrich**
This PR establishes an abstract class `LocalThresholdMethodHistogram` and auto-generated sub-classes that apply the available global threshold methods (e.g. Huang, Otsu, etc) to each neighborhood of the input image individually.

### CODE REVIEW

File: src/main/java/net/imagej/ops/threshold/LocalThresholdMethodHistogram.java [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+			Histogram1d.class, in1() == null ? Iterable.class : in1());
+
+		thresholdComputer = getThresholdComputer();
+	}
+
+	@Override
+	public void compute2(final Iterable<T> neighborhood, final T center,
+		final BitType output)
+	{
+		if (applyThreshold == null) {
```
Comment (dietzc) [inline src/main/java/net/imagej/ops/threshold/LocalThresholdMethodHistogram.java]: initialization not possible in initialize? if not: why? and could we add a `FIXME`. Maybe we can use `NIL` later.

File: src/main/java/net/imagej/ops/threshold/LocalThresholdMethodHistogram.java [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+			applyThreshold = Computers.binary(ops(), Ops.Threshold.Apply.class, output,
+				center, center);
+		}
+		
+		// Compute histogram for neighborhood
+		final Histogram1d<T> hist = histCreator.compute1(neighborhood);
+
+		// Compute threshold
+		T computedThreshold = center.createVariable();
```
Comment (dietzc) [inline src/main/java/net/imagej/ops/threshold/LocalThresholdMethodHistogram.java]: can be final

File: src/main/templates/net/imagej/ops/threshold/ApplyThresholdMethodLocal.vm [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+import net.imagej.ops.threshold.apply.LocalThreshold;
+import net.imglib2.histogram.Histogram1d;
+import net.imglib2.img.Img;
+import net.imglib2.type.logic.BitType;
+import net.imglib2.type.numeric.RealType;
+
+import org.scijava.plugin.Plugin;
+
+/**
+ * Ops that apply a global threshold locally to an {@link Img}.
```
Comment (dietzc) [inline src/main/templates/net/imagej/ops/threshold/ApplyThresholdMethodLocal.vm]: why `Img`?

File: src/main/java/net/imagej/ops/threshold/LocalThresholdMethodHistogram.java [hunk 4]
```diff
@@ -0,0 +1,90 @@
...
+import net.imglib2.type.logic.BitType;
+import net.imglib2.type.numeric.RealType;
+
+/**
+ * {@link AbstractCenterAwareComputerOp} for use in {@link LocalThreshold}s.
+ * 
+ * @author Stefan Helfrich (University of Konstanz)
+ */
+public abstract class LocalThresholdMethodHistogram<T extends RealType<T>>
+	extends AbstractCenterAwareComputerOp<T, BitType>
```
Comment (dietzc) [inline src/main/java/net/imagej/ops/threshold/LocalThresholdMethodHistogram.java]: as it is a computer, we could be more general and say `? extends BooleanType`? Like this we could also pass `BoolType` objects.

File: src/main/java/net/imagej/ops/threshold/LocalThresholdMethodHistogram.java (hunk 1 above)
Comment (stelfrich) [inline src/main/java/net/imagej/ops/threshold/LocalThresholdMethodHistogram.java]: I could move it to initialize with something like this

```applyThreshold = (BinaryComputerOp) Computers.binary(ops(), Ops.Threshold.Apply.class, out()==null?BitType.class:out(), in2()==null?RealType.class:in2(), in2()==null?RealType.class:in2());```

which works but feels really hacky. I have therefore added a `TODO` for moving this code to initialize, once the `NIL` stuff is implemented.

Comment (dietzc) [inline src/main/java/net/imagej/ops/threshold/LocalThresholdMethodHistogram.java]: great, thanks!

File: src/main/java/net/imagej/ops/threshold/LocalThresholdMethodHistogram.java (hunk 4 above)
Comment (stelfrich) [inline src/main/java/net/imagej/ops/threshold/LocalThresholdMethodHistogram.java]: I think it is doable but we currently don't have any obvious ways to create `Img/RAI/...<BoolType>`, or do we @dietzc? So there's not really a chance that a user passes something like that to the `Computer`s, as far as I can tell. If you could tell me how to create an `Img<BoolType>` I could at least write a test to verify that the proposed changes will work out as expected...

Comment (dietzc) [inline src/main/java/net/imagej/ops/threshold/LocalThresholdMethodHistogram.java]: you are right, but this doesn't mean that this won't be possible in the future and there is no disadvantage when we support it, even if we don't test it (correct me if I'm wrong).

Comment (stelfrich) [inline src/main/java/net/imagej/ops/threshold/LocalThresholdMethodHistogram.java]: That's true. And I have implemented some of the changes here, which are compiling. But without being able to create an `Img<BoolType>` in a test, to verify that the implementation really is future-proof, I feel that we should rather not open up the possibility?

Comment (dietzc) [inline src/main/java/net/imagej/ops/threshold/LocalThresholdMethodHistogram.java]: If you implement against `<BooleanType>` and test it against `BitType` that's fine. Then it has to work for `BoolType` as well, right? You also don't test every implementation of `RealType` against each and every `RealType`...

**Comment by stelfrich** [comment]
- [ ] Composition instead of inheritance?

**Comment by ctrueden** [comment]
The merge of this PR broke the `LocalThresholdTest`. At first glance, it seems like the merge itself broke it, as opposed to any commit on the branch—which means it is the interaction of recent changes on master plus this topic branch?

@stelfrich Do you have time to investigate and fix?

**Comment by stelfrich** [comment]
It wasn't your bisect-fu, @ctrueden, the test were passing prior to the merge -- I double checked.

I will investigate immediately.

**Comment by ctrueden** [comment]
Thank you very much!

**Comment by stelfrich** [comment]
@ctrueden I pushed a "fix" to master. We will, however, at some point have to think about how to handle invalid thresholds (which was the reason for the failing tests). I have opened another issue on the topic (#445).

**Pull request merged**